slo-monitor/src/monitors/store.go

```diff
@@ -0,0 +1,196 @@
+	nextIndex int64
+	offset    int64
+}
+
+var keyFunc = cache.DeletionHandlingMetaNamespaceKeyFunc
+
+// NewNoStoreQueue returns newly created NoStoreQueue
+func NewNoStoreQueue() *NoStoreQueue {
+	q := &NoStoreQueue{
+		queue:     []workItem{},
+		indices:   map[string]int64{},
+		nextIndex: 0,
+		offset:    0,
+	}
+
+	q.cond.L = q
+	return q
+}
+
+func (q *NoStoreQueue) addOrModifyItem(key string, obj interface{}, operationType string) {
+	if index, ok := q.indices[key]; ok {
+		q.queue[int(int64(index)-q.offset)].operationType = operationType
+		q.queue[int(int64(index)-q.offset)].obj = obj
+		return
+	}
+	q.indices[key] = q.nextIndex
+	q.nextIndex++
+	q.queue = append(q.queue, workItem{
+		operationType: operationType,
+		key:           key,
+		obj:           obj,
+	})
+}
+
+// Add implements Add function from Store interface
+func (q *NoStoreQueue) Add(obj interface{}) error {
+	key, err := keyFunc(obj)
+	if err != nil {
+		return cache.KeyError{obj, err}
+	}
+
+	q.Lock()
+	defer q.Unlock()
+	defer q.cond.Broadcast()
+	q.addOrModifyItem(key, obj, updateOperationType)
+	return nil
+}
+
+// Update implements Update function from Store interface
+func (q *NoStoreQueue) Update(obj interface{}) error {
+	return q.Add(obj)
+}
+
+// Delete implements Delete function from Store interface
+func (q *NoStoreQueue) Delete(obj interface{}) error {
+	key, err := keyFunc(obj)
+	if err != nil {
+		return cache.KeyError{obj, err}
+	}
+
+	q.Lock()
+	defer q.Unlock()
+	defer q.cond.Broadcast()
+	q.addOrModifyItem(key, obj, deleteOperationType)
+	return nil
+}
+
+// List implements List function from Store interface
+func (q *NoStoreQueue) List() []interface{} {
+	return []interface{}{}
+}
+
+// ListKeys implements ListKeys function from Store interface
+func (q *NoStoreQueue) ListKeys() []string {
+	return []string{}
+}
+
+// Get implements Get function from Store interface
+func (q *NoStoreQueue) Get(obj interface{}) (item interface{}, exists bool, err error) {
+	return "", false, nil
+}
+
+// GetByKey implements GetByKey function from Store interface
+func (q *NoStoreQueue) GetByKey(key string) (item interface{}, exists bool, err error) {
+	return "", false, nil
+}
+
+// Replace implements Replace function from Store interface
+func (q *NoStoreQueue) Replace(list []interface{}, rv string) error {
+	return nil
+}
+
+// Resync implements Resync function from Store interface
+func (q *NoStoreQueue) Resync() error {
+	return nil
+}
+
+// Pop implements Pop function from Queue interface
+func (q *NoStoreQueue) Pop(process cache.PopProcessFunc) (interface{}, error) {
+	q.Lock()
+	defer q.Unlock()
+	for len(q.queue) == 0 {
+		q.cond.Wait()
+	}
+	item := q.queue[0]
+	q.queue = q.queue[1:]
+	q.offset++
+	key, err := keyFunc(item.obj)
+	if err != nil {
+		glog.Errorf("Error in getting key for obj: %v", item.obj)
+	} else {
+		delete(q.indices, key)
+	}
+	err = process(item)
+	if e, ok := err.(cache.ErrRequeue); ok {
+		q.AddIfNotPresent(item)
```

Comment:
This can cause deadlock - as soon as you modify q (and have the item), you should Unlock it.